[AMDGPU][Doc] GFX12.5 Barrier Execution Model (#185632)
llvm/llvm-project: llvm/docs/AMDGPUUsage.rst

  llvm.amdgcn.s.barrier.signal                     Performs a barrier *arrive* operation on the barrier *object* determined by the ``i32`` immediate argument.
                                                   See :ref:`amdgpu-amdhsa-execution-barriers`.
                                                   Available starting GFX12.

  llvm.amdgcn.s.barrier.signal.var                 Performs a barrier *arrive* operation on the barrier *object* determined by the first argument.
                                                   The second argument is an ``i32`` immediate *expected count*. The *expected count* of the barrier *object*
                                                   is only set when the argument is not zero, and when the barrier *object* is a *named barrier object*.
                                                   See :ref:`amdgpu-amdhsa-execution-barriers`.
                                                   Available starting GFX12.

  llvm.amdgcn.s.barrier.signal.isfirst             Performs a barrier *arrive* operation on the barrier *object* determined by the ``i32`` immediate argument.

>                                                    Additionally ensures that the result value is valid even when the intrinsic is used from a wave

Reviewer: What is the result value?

.. note::

  The barrier execution model is experimental and subject to change.

Threads can synchronize execution by performing barrier operations on barrier *objects* as described below:

* Each barrier *object* has the following state:

  * An unsigned positive integer *expected count*: counts the number of *arrive* operations
    expected for this barrier *object*.
  * An unsigned non-negative integer *arrive count*: counts the number of *arrive* operations
    already performed on this barrier *object*.

      * The initial value of *arrive count* is zero.
      * When an operation causes *arrive count* to be equal to *expected count*, the barrier is completed,
        and the *arrive count* is reset to zero.

> * *Barrier-mutually-exclusive* is a symmetric relation between barrier *objects* that represents barrier

Reviewer: ```suggestion
* *Barrier-mutually-exclusive* is a symmetric relation that represents barrier
```
I know the first occurrence gives the type of the relationship and the second occurrence defines it, but it's just redundant when read in plain English.

  * Barrier *wait*.

    * Introduces execution dependencies between threads; this operation depends on
      other barrier operations to complete.

* Barrier modification operations are barrier operations that modify the barrier *object* state:

  * Barrier *init*.
  * Barrier *drop*.
  * Barrier *arrive*.

* *Thread-barrier-order<BO>* is the subset of *program-order* that only
  relates barrier operations performed on a barrier *object* ``BO``.
* All barrier modification operations on a barrier *object* ``BO`` occur in a strict total order called
  *barrier-modification-order<BO>*; It is the order in which ``BO`` observes barrier
  operations that change its state. For any valid *barrier-modification-order<BO>*, the
  following must be true:

  * Let ``A`` and ``B`` be two barrier modification operations where ``A -> B`` in
    *thread-barrier-order<BO>*, then ``A -> B`` is also in *barrier-modification-order<BO>*.
  * The first element in *barrier-modification-order<BO>* is always a barrier *init*, otherwise
    the behavior is undefined.

* *barrier-participates-in* relates barrier operations to the barrier *waits* that depend on them

>   to complete. A barrier operation ``X`` may *barrier-participates-in* a barrier *wait* ``W``

Reviewer: ```suggestion
  to complete. A barrier operation ``X`` *barrier-participates-in* a barrier *wait* ``W``
```
Because the sentence says this is an "if and only if" definition.

  * ``X`` is not a barrier *join*.
  * There is no barrier *join* or *drop* ``JD`` where ``J -> JD -> X`` in *thread-barrier-order<BO>*.
  * There is no barrier *join* ``J'`` on a distinct barrier *object* ``BO'`` such that ``J -> J' -> X`` in
    *program-order*, and ``BO`` *barrier-mutually-exclusive* ``BO'``.

* A barrier operation ``A`` *barrier-executes-before* another barrier operation ``B`` if any of the
  following is true:

  * ``A -> B`` in *program-order*.
  * ``A-> B`` in *barrier-participates-in*.
  * ``A`` *barrier-executes-before* some barrier operation ``X``, and ``X``
    *barrier-executes-before* ``B``.

* *Barrier-executes-before* is consistent with *barrier-modification-order<BO>*
  for every barrier object ``BO``.
* For every barrier *drop* ``D`` performed on a barrier *object* ``BO``:

  * There is a barrier *join* ``J`` such that ``J -> D`` in *barrier-joined-before*;
    otherwise, the behavior is undefined.
  * ``D`` cannot cause the *expected count* of ``BO`` to become negative; otherwise, the behavior is undefined.

* For every pair of barrier *arrive* ``A`` and barrier *drop* ``D`` performed on a barrier *object*

>   ``BO``, such that ``A -> D`` in *thread-barrier-ordered<BO>*, one of the following must be true:

Reviewer: ```suggestion
  ``BO``, such that ``A -> D`` in *thread-barrier-order<BO>*, one of the following must be true:
```

>   * ``A`` does not *barrier-participates-in* any barrier *wait*.

Reviewer: What does that mean? What did A do? Can it exist in an execution that does not trigger undefined behaviour?

  * ``A`` *barrier-participates-in at least one barrier *wait* ``W``
    such that  ``W -> D`` in *barrier-executes-before*.

* For every barrier *wait* ``W`` performed on a barrier *object* ``BO``:

  * There is at least one barrier operation that *barrier-participates-in* ``W``.
  * There is a barrier *join* ``J`` such that ``J -> W`` in *barrier-joined-before*.
  * ``J`` must *barrier-executes-before* at least one operation ``X`` that
    *barrier-participates-in* ``W``; otherwise, the behavior is undefined.

* *barrier-phase-with* is a symmetric relation over barrier operations defined as the
  transitive closure of: *barrier-participates-in* and its inverse relation.
* For every barrier operation ``A`` that *barrier-executes-before* another barrier operation ``B``,
  at least one of the following statements is true; otherwise, the behavior is undefined:

  * ``A`` and ``B`` are performed on different barrier *objects*.
  * ``A`` *barrier-phase-with* ``B``.

>   * Either ``A`` or ``B`` are not *barrier-phase-with* another barrier operation.

Reviewer: Say two threads T1 and T2 synchronize two times over the same barrier, with some intervening synchronization to make sure it is all well-defined. A is in the first synchronization (or phase) and B is in the second phase. Clearly A and B are not related in phase-with. But each of them is in phase-with other operations.

* Each barrier *object* has the following state:

  * An unsigned positive integer *expected count*: counts the number of *arrive* operations
    expected for this barrier *object*.
  * An unsigned non-negative integer *arrive count*: counts the number of *arrive* operations
    already performed on this barrier *object*.

      * The initial value of *arrive count* is zero.
      * When an operation causes *arrive count* to be equal to *expected count*, the barrier is completed,
        and the *arrive count* is reset to zero.

* *Barrier-mutually-exclusive* is a symmetric relation between barrier *objects* that represents barrier
  *objects* that share resources in a way that prevents a thread from using them at the same time.
* Barrier operations are performed on barrier *objects*. A barrier operation is a dynamic instance
  of one of the following:

  * Barrier *init*

    * Barrier *init* takes an additional unsigned positive integer argument *k*.
    * Sets the *expected count* of the *barrier object* to *k*.
    * Resets the *arrive count* of the *barrier object* to zero.

>   * Barrier *join*.

Reviewer: Missing descriptions? Makes this look like fallthrough to the next item.

  * Barrier *drop*.

    * Decrements *expected count* of the barrier *object* by one.

  * Barrier *arrive*.

    * Increments the *arrive count* of the barrier *object* by one.

>     * Depending on the implementation, *arrive* can also update the *expected count* of the

Reviewer: ```suggestion
    * If supported, an additional argument to *arrive* can also be used to update the *expected count* of the
```

      barrier *object* before the *arrive count* is incremented;
      the new *expected count* can never be less than or equal to the *arrive count*,
      otherwise the behavior is undefined.

  * Barrier *wait*.

* Barrier modification operations are barrier operations that modify the barrier *object* state:

  * Barrier *init*.
  * Barrier *drop*.
  * Barrier *arrive*.

* For a given barrier *object* ``BO``, the following relations exist in any
  valid program execution:

>   * *Thread-barrier-order<BO>* is the subset of *program-order* that only

Author: Done

    following must be true:

    * Let ``A`` and ``B`` be two barrier modification operations where ``A -> B`` in
      *thread-barrier-order<BO>*, then ``A -> B`` is also in *barrier-modification-order<BO>*.
    * The first element in *barrier-modification-order<BO>* is always a barrier *init*, otherwise
      the behavior is undefined.

  * *Barrier-participates-in<BO>* relates barrier operations to the barrier *waits* that depend on them
    to complete. A barrier operation ``X`` may *barrier-participates-in<BO>* a barrier *wait* ``W``
    if and only if all of the following is true:

    * ``X`` and ``W`` are both performed on ``BO``.
    * ``X`` is a barrier *arrive* or *drop* operation.
    * ``X`` does not already *barrier-participate-in<BO>* a distinct barrier *wait* ``W'`` in the same thread as ``W``.
    * ``W -> X`` not in *thread-barrier-order<BO>*.
    * All dependent constraint and relations are satisfied as well. [0]_

* Let ``S`` be the set of barrier operations that *barrier-participate-in<BO>* a barrier *wait* ``W`` for some
  barrier *object* ``BO``, then all of the following is true.

>   * The elements of ``S`` all exist in a continuous interval of *barrier-modification-order<BO>*.

Author: Done

  * ``A`` *barrier-participates-in<BO>* at least one barrier *wait* ``W``
    such that  ``W -> D`` in *barrier-executes-before*.

* For every barrier *wait* ``W`` performed on a barrier *object* ``BO``:

  * There is at least one barrier operation that *barrier-participates-in<BO>* ``W``.
  * There is a barrier *join* ``J`` such that ``J -> W`` in *barrier-joined-before*.
  * ``J`` must *barrier-executes-before<BO>* at least one operation ``X`` that
    *barrier-participates-in<BO>* ``W``; otherwise, the behavior is undefined.

* For every barrier *join* ``J`` performed on a barrier *object* ``BO``:

  * ``J`` is not *barrier-joined-before* another barrier *join*.

* A barrier operation ``A`` is *barrier-phase-with* another barrier operation ``B`` if and only if:

  * There exist some barrier *wait* ``W`` such that both ``A -> W`` and ``B -> W`` in
    *barrier-participates-in<BO>* for some barrier *object* ``BO``.

* For every barrier operation ``A`` that *barrier-executes-before* another barrier operation ``B``,
  at least one of the following statements is true; otherwise, the behavior is undefined:

  * ``A`` and ``B`` are performed on different barrier *objects*, or

>   * ``A`` *barrier-phase-with* ``B``, or

Reviewer: But `barrier-executes-before` is a transitive closure over the union of `thread-barrier-order` and `barrier-participates-in`. So in fact it is meant to create a notion of time across all phases? Now I am also wondering if `barrier-executes-before` should be renamed to `barrier-happens-before`.

All barrier *objects* have the following additional target-specific properties:

* Barrier *join* does not increment the *expected count* of a barrier *object*. The *expected count* is set

>   during initialization of the barrier by the hardware.

Author: I just removed the sentence. It was from before, when "drop" was "leave", and this sentence tried to clarify that join and leave are not opposite
The new name is clearer and this sentence is not really useful anymore IMO

* *Barrier-executes-before* relates the dynamic instances of operations from different threads together.
  For example, if ``A -> B`` in *barrier-executes-before*, then the execution of ``A`` must complete
  before the execution of ``B`` can complete.

  * This property can also be combined with *program-order*. For example, let two (non-barrier) operations
    ``X`` and ``Y`` where ``X -> A`` and ``B -> Y`` in *program-order*, then we know that the execution
    of ``X`` completes before the execution of ``Y`` does.

* Barriers do not complete "out-of-thin-air"; a barrier *wait* ``W`` cannot depend on a barrier operation
  ``X`` to complete if ``W -> X`` in *barrier-executes-before*.
* It is undefined behavior to operate on an uninitialized barrier object.
* It is undefined behavior for a barrier *wait* to never complete.
* It is not mandatory to *drop* a barrier after *joining* it. The operations are not opposites; *drop*
  affects future barrier operations by decrementing the *expected count* of the barrier *object*, which

>   can only be undone by re-*initializing* the barrier.

Author: I also removed most of this text for the same reason as above, join/drop is less ambiguous than join/leave and this text caused more confusion than it helped

> * A thread may not *arrive* at then *drop* a barrier *object* unless the barrier completes before the

Reviewer 2: ```suggestion
* A thread may not *arrive* and then *drop* a barrier *object* unless the barrier completes before the
```
